src/feed.sol

```diff
@@ -0,0 +1,148 @@
+
+    // gas-optimized implementation instead of a struct each
+    // variable has its own mapping
+
+    // linked list of future value buckets
+    // normalized timestamp -> value denominated in WAD
+    mapping (uint => uint) public dateBucket;
+    // normalized timestamp -> normalized timestamp
+    mapping (uint => uint) public nextBucket;
+    uint public firstBucket;
+
+    // nftID => maturityDate
+    mapping (bytes32 => uint) public maturityDate;
+
+
+    uint public discountRate;
+    uint public maxDays;
+
+    uint constant NullDate = 1;
+
+    constructor (uint discountRate_, uint maxDays_) public {
+        discountRate = discountRate_;
+        maxDays = maxDays_;
+    }
+
+    function uniqueDayTimestamp(uint timestamp) public pure returns (uint) {
+        return (1 days) * (timestamp/(1 days));
+    }
+
+    /// maturityDate is a unix timestamp
+    function file(bytes32 what, bytes32 nftID_, uint maturityDate_) public {
+        if (what == "maturityDate") {
+            maturityDate[nftID_] = uniqueDayTimestamp(maturityDate_);
+        } else { revert("unknown config parameter");}
+    }
+
+    /// Ceiling Implementation
+    function borrow(uint loan, uint amount) external auth {
+        uint normalizedDay = uniqueDayTimestamp(now);
+
+        // ceiling check uses existing loan debt
+        require(ceiling(loan) >= safeAdd(pile.debt(loan), amount), "borrow-amount-too-high");
+
+        bytes32 nftID_ = nftID(loan);
+        // calculate future cash flow
+        uint maturityDate_ = maturityDate[nftID_];
+
+
+        if (dateBucket[maturityDate_] == 0) {
+            addToLinkedList(maturityDate_);
+        }
+
+        // calculate future value of the loan and add it to the bucket
+        dateBucket[maturityDate_] = safeAdd(dateBucket[maturityDate_],
+            rmul(rpow(pile.loanRates(loan),  safeSub(maturityDate_, normalizedDay), ONE), amount));
+
+    }
+
+    function addToLinkedList(uint maturityDate_) internal {
+        if (firstBucket == 0) {
+            firstBucket = maturityDate_;
+            nextBucket[maturityDate_] = NullDate;
+            return;
+        }
+
+        // new bucket before first one
+        if (maturityDate_ < firstBucket) {
+            nextBucket[maturityDate_] = firstBucket;
+            firstBucket = maturityDate_;
+            return;
+        }
+
+        // find previous bucket
+        uint prev = maturityDate_;
+        while(nextBucket[prev] == 0) {prev = prev - 1 days;}
+
+        nextBucket[maturityDate_] = nextBucket[prev];
+        nextBucket[prev] = maturityDate_;
+    }
+
+    function repay(uint loan, uint amount) external auth {
+        // todo
+        // remove from FV
+        // remove from linked list if bucket is zero
+
+    }
+
+    /// deprecated
+    /// returns the NAV (net asset value) of the pool
+    /// only for performance comparing
+    /// todo old implementation will be removed
+    function navOld() public view returns(uint) {
```

Comment:
only here to compare the performance during development, so far the linked list has a way better runtime.